snakemake/script.py

```diff
@@ -176,71 +259,510 @@ def __init__(
+        )
 
-        This function allows scripts and wrappers support optional `log` files
-        specified in the calling rule.  If no `log` was specified, then an
-        empty string "" is returned, regardless of the values of `stdout`,
-        `stderr`, and `append`.
+    def get_preamble(self):
+        wrapper_path = self.path[7:] if self.path.startswith("file://") else self.path
+        preamble_addendum = "__real_file__ = __file__; __file__ = {file_override};".format(
+            file_override=repr(os.path.realpath(wrapper_path))
+        )
 
-        Parameters
-        ---------
+        return PythonScript.generate_preamble(
+            self.path,
+            self.source,
+            self.basedir,
+            self.input,
+            self.output,
+            self.params,
+            self.wildcards,
+            self.threads,
+            self.resources,
+            self.log,
+            self.config,
+            self.rulename,
+            self.conda_env,
+            self.singularity_img,
+            self.singularity_args,
+            self.bench_record,
+            self.jobid,
+            self.bench_iteration,
+            self.cleanup_scripts,
+            self.shadow_dir,
+            preamble_addendum=preamble_addendum,
+        )
 
-        stdout : bool
-            Send stdout to log
+    def write_script(self, preamble, fd):
+        fd.write(preamble.encode())
+        fd.write(self.source)
+
+    def execute_script(self, fname):
+        py_exec = sys.executable
+        if self.conda_env is not None:
+            py = os.path.join(self.conda_env, "bin", "python")
+            if os.path.exists(py):
+                out = subprocess.check_output(
+                    [py, "--version"],
+                    stderr=subprocess.STDOUT,
+                    universal_newlines=True,
+                )
+                ver = tuple(map(int, PY_VER_RE.match(out).group("ver_min").split(".")))
+                if ver >= MIN_PY_VERSION:
+                    # Python version is new enough, make use of environment
+                    # to execute script
+                    py_exec = "python"
+                else:
+                    logger.warning(
+                        "Conda environment defines Python "
+                        "version < {0}.{1}. Using Python of the "
+                        "master process to execute "
+                        "script. Note that this cannot be avoided, "
+                        "because the script uses data structures from "
+                        "Snakemake which are Python >={0}.{1} "
+                        "only.".format(*MIN_PY_VERSION)
+                    )
+        if self.singularity_img is not None:
+            # use python from image
+            py_exec = "python"
+        # use the same Python as the running process or the one from the environment
+        shell("{py_exec} {fname:q}", bench_record=self.bench_record)
```

Comment:
Ok, I see the problem that manifests with conda_env for you. Here, `conda_env` and `singularity_img` must be available as local variables (this is where shell searches them).

Reply:
I guess it makes sense to have a separate method in the ABC for executing the command, which handles this for all inheriting classes.

Reply:
The following will work once you have merged in the latest master.
```
    def execute_cmd(self, cmd):
        shell("{cmd}, bench_record=self.bench_record, conda_env=self.conda_env, singularity_img=self.singularity_img, shadow_dir=self.shadow_dir, env_modules=self.env_modules)

Reply:
Ok, master is merged in. Let's see.

Reply:
Yes, this generally makes sense.
However, if we introduce the function like this:
```python
    def _execute_cmd(self, cmd, **kwargs):
        shell(
            "{cmd}",
            bench_record=self.bench_record,
            conda_env=self.conda_env,
            singularity_img=self.singularity_img,
            shadow_dir=self.shadow_dir,
            env_modules=self.env_modules,
            **kwargs)
```
then this method will no longer have e.g. `py_exec` as its local variable (since we are one function call deeper).

Reply:
I now implemented it like this:
```python
    def _execute_cmd(self, cmd, **kwargs):
        context = inspect.currentframe().f_back.f_locals
        context.update(kwargs)

        shell(
            cmd,
            bench_record=self.bench_record,
            conda_env=self.conda_env,
            singularity_img=self.singularity_img,
            shadow_dir=self.shadow_dir,
            env_modules=self.env_modules,
            **context)
```
(see 9579090ea8f62b06679b8cbb3cdaf9738bc3accc)

It seems to work but is a bit convoluted.

Reply:
I don't think we need context here. You should be able to simply pass py_exec to _execute_cmd as a keyword argument, shouldn't you?

Reply:
That will certainly also work.

It would then be necessary to pass all used variables to `self._execute_cmd(..)`. See e.g.:
```python
self._execute_cmd("{py_exec} {fname:q}", py_exec=py_exec, fname=fname)
# instead of
self._execute_cmd("{py_exec} {fname:q}")
```

This would be fine for me (I think I even prefer it, because it is more explicit).
So far I was mainly trying to replicate the way `shell(..)` works.

Reply:
Exactly, I prefer it for the same reason. Shell just works like that for convenience when writing a workflow. For internal use, it should always be explicit over implcit.